### PR TITLE
Validate documents 

### PR DESCRIPTION
**Pull request by guilload**
### Description
Parse and validate documents before persisting them into WAL. Returns parse failures to clients.

### How was this PR tested?
Added unit tests.

### CODE REVIEW

File: quickwit/quickwit-ingest/src/ingest_v2/mod.rs [hunk 1]
```diff
@@ -145,20 +162,20 @@ pub struct IngestRequestV2Builder {
 
 impl IngestRequestV2Builder {
     /// Adds a document to the request.
```
Comment (fulmicoton) [inline quickwit/quickwit-ingest/src/ingest_v2/mod.rs:164]: Can you comment what the returned value is about?

File: quickwit/quickwit-proto/src/types/doc_uid.rs [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+        #[allow(clippy::unwrap_or_default)]
+        let next_ulid = self.0.increment().unwrap_or_else(Ulid::new);
+        Self(next_ulid)
+    }
+}
+
+impl<'de> Deserialize<'de> for DocUid {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where D: Deserializer<'de> {
+        let ulid_str: String = String::deserialize(deserializer)?;
```
Comment (fulmicoton) [inline quickwit/quickwit-proto/src/types/doc_uid.rs:74]: I think the following works.
(I have no clue if we call it a lot or though)

```suggestion
        let ulid_str: Cow<'de, str> = Cow::deserialize(deserializer)?;
```

File: quickwit/quickwit-proto/src/types/doc_uid.rs [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+        let ulid = Ulid::from_string(&ulid_str).map_err(D::Error::custom)?;
+        Ok(Self(ulid))
+    }
+}
+
+impl Serialize for DocUid {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where S: Serializer {
```
Comment (fulmicoton) [inline quickwit/quickwit-proto/src/types/doc_uid.rs:82]: can we have a dumb unit test for serialize / deserialize?

File: quickwit/quickwit-proto/src/types/doc_uid.rs [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+        &mut self,
+        tag: u32,
+        wire_type: prost::encoding::WireType,
+        buf: &mut B,
+        ctx: prost::encoding::DecodeContext,
+    ) -> ::core::result::Result<(), prost::DecodeError>
+    where
+        B: prost::bytes::Buf,
+    {
+        const STRUCT_NAME: &str = "DocUid";
```
Comment (fulmicoton) [inline quickwit/quickwit-proto/src/types/doc_uid.rs:104]: Same as above, a unit test would be useful here.

File: quickwit/quickwit-serve/Cargo.toml [hunk 5]
```diff
@@ -44,6 +44,7 @@ tokio-stream = { workspace = true }
 tower = { workspace = true }
 tower-http = { workspace = true }
 tracing = { workspace = true }
+ulid = { workspace = true }
```
Comment (fulmicoton) [inline quickwit/quickwit-serve/Cargo.toml:47]:
```suggestion
```

File: quickwit/rest-api-tests/scenarii/es_compatibility/bulk/_teardown.elasticsearch.yaml [hunk 6]
```diff
@@ -1,3 +0,0 @@
-method: DELETE
```
Comment (fulmicoton) [inline quickwit/rest-api-tests/scenarii/es_compatibility/bulk/_teardown.elasticsearch.yaml:1]: I assume you removed it for debugging. Please leave it though.

File: quickwit/quickwit-serve/src/elasticsearch_api/bulk_v2.rs [hunk 7]
```diff
@@ -148,19 +165,69 @@ pub(crate) async fn elastic_bulk_ingest_v2(
...
+        let index_id = success
+            .index_uid
+            .map(|index_uid| index_uid.index_id)
+            .expect("`index_uid` should be a required field");
+
+        // Populate the response items with one error per parse failure.
+        for parse_failure in success.parse_failures {
+            // Since the generated doc UIDs are monotonically increasing, and inserted in order, we
+            // can find doc handles using binary search.
+            assert!(
```
Comment (fulmicoton) [inline quickwit/quickwit-serve/src/elasticsearch_api/bulk_v2.rs:177]: Let's move this assert out of the loop?

File: quickwit/quickwit-serve/src/elasticsearch_api/bulk_v2.rs [hunk 8]
```diff
@@ -148,19 +165,69 @@ pub(crate) async fn elastic_bulk_ingest_v2(
...
+                doc_handles
+                    .windows(2)
+                    .all(|doc_handles| doc_handles[0].doc_uid < doc_handles[1].doc_uid),
+                "doc handles should be sorted by doc UID"
+            );
+            let doc_handle_idx = doc_handles
+                .binary_search_by(|doc_handle| doc_handle.doc_uid.cmp(&parse_failure.doc_uid()))
```
Comment (fulmicoton) [inline quickwit/quickwit-serve/src/elasticsearch_api/bulk_v2.rs:184]: nitpick
I think with `Copy` stuff you can use binary_search_by_key`.

Comment (guilload) [reply quickwit/quickwit-serve/src/elasticsearch_api/bulk_v2.rs:184]: good point.

File: quickwit/quickwit-serve/src/elasticsearch_api/bulk_v2.rs (hunk 7 above)
Comment (guilload) [reply quickwit/quickwit-serve/src/elasticsearch_api/bulk_v2.rs:177]: No, the `Vec` of `DocHandle` is built locally and doesn't go through gRPC.

**Review by fulmicoton — APPROVED** [review]

File: quickwit/quickwit-proto/src/types/doc_uid.rs (hunk 3 above)
Comment (guilload) [reply quickwit/quickwit-proto/src/types/doc_uid.rs:82]: Yes, I was working on adding those.

File: quickwit/quickwit-proto/src/types/doc_uid.rs (hunk 2 above)
Comment (guilload) [reply quickwit/quickwit-proto/src/types/doc_uid.rs:74]: Yes, it works. There are a few other places that could benefit from this pattern.

**Review by fulmicoton — APPROVED** [review]

File: quickwit/quickwit-serve/src/elasticsearch_api/bulk_v2.rs [hunk 9]
```diff
@@ -24,14 +24,14 @@ use hyper::StatusCode;
 use quickwit_config::INGEST_V2_SOURCE_ID;
```
Comment (guilload) [inline quickwit/quickwit-serve/src/elasticsearch_api/bulk_v2.rs:24]: TODO @guilload: Add some unit tests for this file, and then the PR is ready.

File: quickwit/quickwit-serve/src/elasticsearch_api/bulk_v2.rs [hunk 10]
```diff
@@ -111,15 +120,20 @@ pub(crate) async fn elastic_bulk_ingest_v2(
...
             })?;
-        let subrequest_id = ingest_request_builder.add_doc(index_id, source);
-
-        per_subrequest_id_es_doc_ids
+        let (subrequest_id, doc_uid) = ingest_request_builder.add_doc(index_id, doc);
+
+        let doc_handle = DocHandle {
+            doc_uid,
+            es_doc_id: meta.es_doc_id,
+            successfully_parsed: true,
```
[the remark this answers is not in the document]
Comment (guilload) [reply quickwit/quickwit-serve/src/elasticsearch_api/bulk_v2.rs]: replaced with `is_parse_failure` for clarity.

**Pull request merged**